source/m2mfirmware.cpp

```diff
@@ -55,35 +55,125 @@ M2MFirmware::~M2MFirmware()
+    (uint8_t*)"1",          // path
+    (uint8_t*)"",           // resource
+    1,                      // strlen("1")
+    0,                      // resourcelen
+    false,                  // external_memory_block
+    SN_GRS_DYNAMIC,         // mode
+    false                   // free_on_delete
+};
+
+static sn_nsdl_static_resource_parameters_s firmware_update_params_static = {
+    (char*)OMA_RESOURCE_TYPE,      // resource_type_ptr
+    (char*)"",                     // interface_description_ptr
+    (uint8_t*)"2",          // path
+    (uint8_t*)"",           // resource
+    1,                      // strlen("2")
+    0,                      // resourcelen
+    false,                  // external_memory_block
+    SN_GRS_DYNAMIC,         // mode
+    false                   // free_on_delete
+};
+
+
+static sn_nsdl_dynamic_resource_parameters_s firmware_package_params_dynamic = {
+    NULL,
+    &firmware_package_params_static,
+    {NULL, NULL},                     // link
+    COAP_CONTENT_OMA_TLV_TYPE,        // coap_content_type
+    M2MBase::PUT_ALLOWED,   // access
+    0,                      // registered
+    false,                  // publish_uri
+    false,                  // free_on_delete
+    true                    // observable
+};
+
+static sn_nsdl_dynamic_resource_parameters_s firmware_package_uri_params_dynamic = {
+    NULL,
+    &firmware_package_uri_params_static,
+    {NULL, NULL},                     // link
+    COAP_CONTENT_OMA_TLV_TYPE,        // coap_content_type
+    M2MBase::PUT_ALLOWED,   // access
+    0,                      // registered
+    false,                  // publish_uri
+    false,                  // free_on_delete
+    true                    // observable
+};
+
+static sn_nsdl_dynamic_resource_parameters_s firmware_update_params_dynamic = {
+    NULL,
+    &firmware_update_params_static,
+    {NULL, NULL},                     // link
+    COAP_CONTENT_OMA_TLV_TYPE,        // coap_content_type
+    M2MBase::NOT_ALLOWED,   // access
+    0,                      // registered
+    false,                  // publish_uri
+    false,                  // free_on_delete
+    true                    // observable
+};
+
+
+const static M2MBase::lwm2m_parameters firmware_package_params = {
+    0, // max_age
+    0, // instance_id
+    0, // name_id
+    FIRMWARE_PACKAGE, // name
+    &firmware_package_params_dynamic,
+    M2MBase::Resource // base_type
+};
+
+const static M2MBase::lwm2m_parameters firmware_package_uri_params = {
+    0, // max_age
+    0, // instance_id
+    0, // name_id
+    FIRMWARE_PACKAGE_URI, // name
+    &firmware_package_uri_params_dynamic,
+    M2MBase::Resource // base_type
+};
+
+const static M2MBase::lwm2m_parameters firmware_update_params = {
+    0, // max_age
+    0, // instance_id
+    0, // name_id
+    FIRMWARE_UPDATE, // name
+    &firmware_update_params_dynamic,
+    M2MBase::Resource // base_type
+};
+
```

Comment:
Content type should be changed from tlv to text/plain. TLV is applicable only in object/object instance level or in case resource has several instances.